# Projeto de processamento de dados e análise exploratória


* **Importando as bibliotecas**

In [2]:
import pandas as pd
import numpy as np

* **Leitura dos dados de entrada**

In [3]:
df_movies = pd.read_csv('movies.csv', sep=',')
df_ratings = pd.read_csv('ratings.csv', sep=',')

In [4]:
df_movies

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

In [5]:
df_ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]

* **Mesclando os dados**

In [6]:
df_movie_lens = pd.merge(df_ratings, df_movies, how ='left', on = 'movieId')

In [7]:
df_movie_lens.head()

userId  movieId  rating  timestamp                        title  \
0       1        1     4.0  964982703             Toy Story (1995)   
1       1        3     4.0  964981247      Grumpier Old Men (1995)   
2       1        6     4.0  964982224                  Heat (1995)   
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  964982931   Usual Suspects, The (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

* **Excluindo coluna 'timestamp' que não vai ser necessária**

In [8]:
df_movie_lens.pop('timestamp')

0          964982703
1          964981247
2          964982224
3          964983815
4          964982931
             ...    
100831    1493848402
100832    1493850091
100833    1494273047
100834    1493846352
100835    1493846415
Name: timestamp, Length: 100836, dtype: int64

* **Renomeando as colunas**

In [9]:
df_movie_lens.rename(columns={'userId': 'usuario_id', 'movieId': 'filmes_id', 'rating': 'nota', 'title': 'titulo', 'genres': 'genero'}, inplace = True)

* **Visualizando o novo DataFrame**

In [10]:
df_movie_lens.head()

usuario_id  filmes_id  nota                       titulo  \
0           1          1   4.0             Toy Story (1995)   
1           1          3   4.0      Grumpier Old Men (1995)   
2           1          6   4.0                  Heat (1995)   
3           1         47   5.0  Seven (a.k.a. Se7en) (1995)   
4           1         50   5.0   Usual Suspects, The (1995)   

                                        genero  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

* **Resumo de valores não nulos**

In [11]:
df_movie_lens.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100836 entries, 0 to 100835
Data columns (total 5 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   usuario_id  100836 non-null  int64  
 1   filmes_id   100836 non-null  int64  
 2   nota        100836 non-null  float64
 3   titulo      100836 non-null  object 
 4   genero      100836 non-null  object 
dtypes: float64(1), int64(2), object(2)
memory usage: 4.6+ MB


* **Resumo estatístico das variáveis numéricas**

In [12]:
df_movie_lens.describe()

usuario_id      filmes_id           nota
count  100836.000000  100836.000000  100836.000000
mean      326.127564   19435.295718       3.501557
std       182.618491   35530.987199       1.042529
min         1.000000       1.000000       0.500000
25%       177.000000    1199.000000       3.000000
50%       325.000000    2991.000000       3.500000
75%       477.000000    8122.000000       4.000000
max       610.000000  193609.000000       5.000000

# **Começando as análises**

In [13]:
df_movie_lens.head()

usuario_id  filmes_id  nota                       titulo  \
0           1          1   4.0             Toy Story (1995)   
1           1          3   4.0      Grumpier Old Men (1995)   
2           1          6   4.0                  Heat (1995)   
3           1         47   5.0  Seven (a.k.a. Se7en) (1995)   
4           1         50   5.0   Usual Suspects, The (1995)   

                                        genero  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

* **Quantos gêneros o filme "The Matrix" tem? Quais são os gêneros?**

In [14]:
df_movie_lens[df_movie_lens['titulo'] == 'Matrix, The (1999)']['genero'].unique()

array(['Action|Sci-Fi|Thriller'], dtype=object)

O filme **The Matrix** tem 3 gêneros, que são eles: **Action, Sci-Fi, Thriller**

* **Quantos gêneros cada filme tem, em média?**

In [15]:
#criando uma lista de genero e adicionando na coluna 'generos'
df_movie_lens['generos'] = df_movie_lens['genero'].str.split('|')

In [16]:
df_movie_lens

usuario_id  filmes_id  nota                          titulo  \
0                1          1   4.0                Toy Story (1995)   
1                1          3   4.0         Grumpier Old Men (1995)   
2                1          6   4.0                     Heat (1995)   
3                1         47   5.0     Seven (a.k.a. Se7en) (1995)   
4                1         50   5.0      Usual Suspects, The (1995)   
...            ...        ...   ...                             ...   
100831         610     166534   4.0                    Split (2017)   
100832         610     168248   5.0   John Wick: Chapter Two (2017)   
100833         610     168250   5.0                  Get Out (2017)   
100834         610     168252   5.0                    Logan (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   

                                             genero  \
0       Adventure|Animation|Children|Comedy|Fantasy   
1                                    Comedy|Romance   
2                             Action|Crime|Thriller   
3                                  Mystery|Thriller   
4                            Crime|Mystery|Thriller   
...                                             ...   
100831                        Drama|Horror|Thriller   
100832                        Action|Crime|Thriller   
100833                                       Horror   
100834                                Action|Sci-Fi   
100835                  Action|Crime|Drama|Thriller   

                                                  generos  
0       [Adventure, Animation, Children, Comedy, Fantasy]  
1                                       [Comedy, Romance]  
2                               [Action, Crime, Thriller]  
3                                     [Mystery, Thriller]  
4                              [Crime, Mystery, Thriller]  
...                                                   ...  
100831                          [Drama, Horror, Thriller]  
100832                          [Action, Crime, Thriller]  
100833                                           [Horror]  
100834                                   [Action, Sci-Fi]  
100835                   [Action, Crime, Drama, Thriller]  

[100836 rows x 6 columns]

In [17]:
#excluindo coluna 'genero' que não vai ser usada
df_movie_lens.pop('genero')

0         Adventure|Animation|Children|Comedy|Fantasy
1                                      Comedy|Romance
2                               Action|Crime|Thriller
3                                    Mystery|Thriller
4                              Crime|Mystery|Thriller
                             ...                     
100831                          Drama|Horror|Thriller
100832                          Action|Crime|Thriller
100833                                         Horror
100834                                  Action|Sci-Fi
100835                    Action|Crime|Drama|Thriller
Name: genero, Length: 100836, dtype: object

In [18]:
df_movie_lens

usuario_id  filmes_id  nota                          titulo  \
0                1          1   4.0                Toy Story (1995)   
1                1          3   4.0         Grumpier Old Men (1995)   
2                1          6   4.0                     Heat (1995)   
3                1         47   5.0     Seven (a.k.a. Se7en) (1995)   
4                1         50   5.0      Usual Suspects, The (1995)   
...            ...        ...   ...                             ...   
100831         610     166534   4.0                    Split (2017)   
100832         610     168248   5.0   John Wick: Chapter Two (2017)   
100833         610     168250   5.0                  Get Out (2017)   
100834         610     168252   5.0                    Logan (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   

                                                  generos  
0       [Adventure, Animation, Children, Comedy, Fantasy]  
1                                       [Comedy, Romance]  
2                               [Action, Crime, Thriller]  
3                                     [Mystery, Thriller]  
4                              [Crime, Mystery, Thriller]  
...                                                   ...  
100831                          [Drama, Horror, Thriller]  
100832                          [Action, Crime, Thriller]  
100833                                           [Horror]  
100834                                   [Action, Sci-Fi]  
100835                   [Action, Crime, Drama, Thriller]  

[100836 rows x 5 columns]

In [19]:
#criando uma nova coluna 'n_generos' (conta generos)
df_movie_lens['n_generos'] = [len(x) for x in df_movie_lens['generos']]

In [20]:
df_movie_lens

usuario_id  filmes_id  nota                          titulo  \
0                1          1   4.0                Toy Story (1995)   
1                1          3   4.0         Grumpier Old Men (1995)   
2                1          6   4.0                     Heat (1995)   
3                1         47   5.0     Seven (a.k.a. Se7en) (1995)   
4                1         50   5.0      Usual Suspects, The (1995)   
...            ...        ...   ...                             ...   
100831         610     166534   4.0                    Split (2017)   
100832         610     168248   5.0   John Wick: Chapter Two (2017)   
100833         610     168250   5.0                  Get Out (2017)   
100834         610     168252   5.0                    Logan (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   

                                                  generos  n_generos  
0       [Adventure, Animation, Children, Comedy, Fantasy]          5  
1                                       [Comedy, Romance]          2  
2                               [Action, Crime, Thriller]          3  
3                                     [Mystery, Thriller]          2  
4                              [Crime, Mystery, Thriller]          3  
...                                                   ...        ...  
100831                          [Drama, Horror, Thriller]          3  
100832                          [Action, Crime, Thriller]          3  
100833                                           [Horror]          1  
100834                                   [Action, Sci-Fi]          2  
100835                   [Action, Crime, Drama, Thriller]          4  

[100836 rows x 6 columns]

In [21]:
#tirando a média da quantidade de generos por filme (e arredondando)
round(df_movie_lens['n_generos'].mean())

3

A **média de generos** por filme é: **3**

* **Quais são os cinco filmes com maior número de gêneros?**

In [22]:
#criando uma variável do dataframe, onde mostra os 5 filmes com maior número de generos
df_cinco_maiores = df_movie_lens.nlargest(5, "n_generos")

In [23]:
df_cinco_maiores

usuario_id  filmes_id  nota  \
29899          204      81132   3.5   
100469         610      81132   4.0   
94392          599      26701   3.0   
194              1       2987   5.0   
246              2      79132   4.0   

                                                   titulo  \
29899                                       Rubber (2010)   
100469                                      Rubber (2010)   
94392   Patlabor: The Movie (Kidô keisatsu patorebâ: T...   
194                       Who Framed Roger Rabbit? (1988)   
246                                      Inception (2010)   

                                                  generos  n_generos  
29899   [Action, Adventure, Comedy, Crime, Drama, Film...         10  
100469  [Action, Adventure, Comedy, Crime, Drama, Film...         10  
94392   [Action, Animation, Crime, Drama, Film-Noir, M...          8  
194     [Adventure, Animation, Children, Comedy, Crime...          7  
246     [Action, Crime, Drama, Mystery, Sci-Fi, Thrill...          7

Os **cinco filmes** com maior número de gêneros são:

In [24]:
df_cinco_maiores['titulo']

29899                                         Rubber (2010)
100469                                        Rubber (2010)
94392     Patlabor: The Movie (Kidô keisatsu patorebâ: T...
194                         Who Framed Roger Rabbit? (1988)
246                                        Inception (2010)
Name: titulo, dtype: object

* Quantos **usuários** deram nota para o filme **"The Matrix"**?

In [25]:
len(df_movie_lens[df_movie_lens['titulo'] == 'Matrix, The (1999)']['usuario_id'])

278

O número de usuários que deram nota pro filme **"The Matrix"** é: **278**

* **Quantos usuários, em média, deram nota para cada filme?**

In [26]:
round(df_movie_lens['usuario_id'].value_counts().mean())

165

Em média, **165** usuários deram nota para cada filme

* **Quais são os cinco filmes para os quais mais usuários deram nota?**

In [27]:
#selecionando os 5 filmes
quantidade_notas_filme = df_movie_lens['titulo'].value_counts()[:5]
quantidade_notas_filme

Forrest Gump (1994)                 329
Shawshank Redemption, The (1994)    317
Pulp Fiction (1994)                 307
Silence of the Lambs, The (1991)    279
Matrix, The (1999)                  278
Name: titulo, dtype: int64

* **Qual a nota média do filme "The Matrix"?**

In [31]:
#pesquisando a nota média do filme "The Matrix" e arredondando
round(df_movie_lens.loc[df_movie_lens['titulo']== 'Matrix, The (1999)'].nota.mean())

4

A **média** de notas do filme **"The Matrix"** é: **4**

* **Cada filme tem uma nota média. Qual a média dessas notas, por gênero?**

In [32]:
df_movie_lens

usuario_id  filmes_id  nota                          titulo  \
0                1          1   4.0                Toy Story (1995)   
1                1          3   4.0         Grumpier Old Men (1995)   
2                1          6   4.0                     Heat (1995)   
3                1         47   5.0     Seven (a.k.a. Se7en) (1995)   
4                1         50   5.0      Usual Suspects, The (1995)   
...            ...        ...   ...                             ...   
100831         610     166534   4.0                    Split (2017)   
100832         610     168248   5.0   John Wick: Chapter Two (2017)   
100833         610     168250   5.0                  Get Out (2017)   
100834         610     168252   5.0                    Logan (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   

                                                  generos  n_generos  
0       [Adventure, Animation, Children, Comedy, Fantasy]          5  
1                                       [Comedy, Romance]          2  
2                               [Action, Crime, Thriller]          3  
3                                     [Mystery, Thriller]          2  
4                              [Crime, Mystery, Thriller]          3  
...                                                   ...        ...  
100831                          [Drama, Horror, Thriller]          3  
100832                          [Action, Crime, Thriller]          3  
100833                                           [Horror]          1  
100834                                   [Action, Sci-Fi]          2  
100835                   [Action, Crime, Drama, Thriller]          4  

[100836 rows x 6 columns]

In [44]:
#criando um novo DataFrame transformando cada genero de 'generos' em uma linha individual
new_df_movie_lens = df_movie_lens.explode('generos')

In [45]:
new_df_movie_lens

usuario_id  filmes_id  nota                          titulo  \
0                1          1   4.0                Toy Story (1995)   
0                1          1   4.0                Toy Story (1995)   
0                1          1   4.0                Toy Story (1995)   
0                1          1   4.0                Toy Story (1995)   
0                1          1   4.0                Toy Story (1995)   
...            ...        ...   ...                             ...   
100834         610     168252   5.0                    Logan (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   
100835         610     170875   3.0  The Fate of the Furious (2017)   

          generos  n_generos  
0       Adventure          5  
0       Animation          5  
0        Children          5  
0          Comedy          5  
0         Fantasy          5  
...           ...        ...  
100834     Sci-Fi          2  
100835     Action          4  
100835      Crime          4  
100835      Drama          4  
100835   Thriller          4  

[274480 rows x 6 columns]

In [46]:
media_generos = new_df_movie_lens.groupby('generos')['nota'].mean()

A nota **média** por **gênero** é: 

In [47]:
media_generos = new_df_movie_lens.groupby('generos')['nota'].mean().to_frame('media')
media_generos.sort_values('media',ascending=True)

media
generos                     
Horror              3.258195
Comedy              3.384721
Children            3.412956
Action              3.447984
Sci-Fi              3.455721
(no genres listed)  3.489362
Fantasy             3.491001
Thriller            3.493706
Romance             3.506511
Adventure           3.508609
Musical             3.563678
Western             3.583938
IMAX                3.618335
Animation           3.629937
Mystery             3.632460
Drama               3.656184
Crime               3.658294
Documentary         3.797785
War                 3.808294
Film-Noir           3.920115

* **Para cada gênero de filme, quantos usuários deram cada nota possível?**

In [42]:
nota_possivel_genero = new_df_movie_lens.groupby('generos')['usuario_id'].count().to_frame('nota_possivel_genero')

A **quantidade** de usuários que deram cada nota possível por **gênero** é: 

In [43]:
nota_possivel_genero

nota_possivel_genero
generos                                 
(no genres listed)                    47
Action                             30635
Adventure                          24161
Animation                           6988
Children                            9208
Comedy                             39053
Crime                              16681
Documentary                         1219
Drama                              41928
Fantasy                            11834
Film-Noir                            870
Horror                              7291
IMAX                                4145
Musical                             4138
Mystery                             7674
Romance                            18124
Sci-Fi                             17243
Thriller                           26452
War                                 4859
Western                             1930